linmod/eval.py

```diff
@@ -1,124 +1,210 @@
+import numpy as np
 import polars as pl
 
 from linmod.utils import pl_list_cycle, pl_norm
 
 
-def _merge_samples_and_data(samples, data):
-    r"""
-    Join the forecast samples and raw data dataframes, assuming they have the
-    standard format.
-
-    Also compute the true proportions from the raw data.
-    """
-    result = (
-        data.with_columns(
-            phi=(
-                pl.col("count") / pl.sum("count").over("fd_offset", "division")
-            ),
-        )
-        .drop("count")
-        .join(
-            samples,
-            on=("fd_offset", "division", "lineage"),
-            how="left",
-            suffix="_sampled",
+def multinomial_count_sampler(N, phi, rng):
+    return rng.multinomial(N, phi)
+
+
+class ProportionsEvaluator:
+    def __init__(self, samples: pl.LazyFrame, data: pl.LazyFrame):
+        """
+        `samples` should have the standard model output format.
+        `data` should have the standard model input format.
+        """
+
+        # Join the forecast samples and raw data dataframes.
+        # Also compute the true proportions from the raw data.
+        self.df = (
+            data.with_columns(
+                phi=(
+                    pl.col("count")
+                    / pl.sum("count").over("fd_offset", "division")
+                ),
+            )
+            .drop("count")
+            # NaN will appear on division-days where none of the lineages
+            # were sampled. We will filter these out.
+            # TODO: Is this good to do?
+            .filter(pl.col("phi").is_not_nan())
+            .join(
+                samples,
+                on=("fd_offset", "division", "lineage"),
+                how="left",
+                suffix="_sampled",
+            )
         )
-    )
-
-    return result
-
 
-def proportions_mean_norm_per_division_day(samples, data, p=1):
-    r"""
-    The expected norm of proportion forecast error for each division-day.
+    def _mean_norm_per_division_day(self, p=1) -> pl.LazyFrame:
+        r"""
```

Comment:
```suggestion
        """
```
The `r` is for regex?

Reply:
oh, this is so you can write latex with backslashes?

personally I'd suggest just removing the backslashes; no one is confused that `phi` means $\phi$

unless this is something about how the docstrings render in mkdocs?

I'm learning something tdoay

Reply:
I like the way it renders in the docs personally :)